timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,378 @@
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pymbar
+from rdkit import Chem
+from rdkit.Chem import AllChem, Draw
+
+from timemachine.constants import BOLTZ
+from timemachine.fe.single_topology_v3 import SingleTopologyV3
+from timemachine.fe.system import convert_bps_into_system
+from timemachine.fe.utils import get_romol_conf
+from timemachine.ff.handlers import openmm_deserializer
+from timemachine.lib import LangevinIntegrator, MonteCarloBarostat, custom_ops
+from timemachine.md import minimizer
+from timemachine.md.barostat.utils import get_bond_list, get_group_indices
+
+
+def plot_atom_mapping_grid(mol_a, mol_b, core_smarts, core, show_idxs=False):
+    mol_a_2d = Chem.Mol(mol_a)
+    mol_b_2d = Chem.Mol(mol_b)
+    mol_q_2d = Chem.MolFromSmarts(core_smarts)
+
+    AllChem.Compute2DCoords(mol_q_2d)
+
+    q_to_a = [[int(x[0]), int(x[1])] for x in enumerate(core[:, 0])]
+    q_to_b = [[int(x[0]), int(x[1])] for x in enumerate(core[:, 1])]
+
+    AllChem.GenerateDepictionMatching2DStructure(mol_a_2d, mol_q_2d, atomMap=q_to_a)
+    AllChem.GenerateDepictionMatching2DStructure(mol_b_2d, mol_q_2d, atomMap=q_to_b)
+
+    atom_colors_a = {}
+    atom_colors_b = {}
+    atom_colors_q = {}
+    for c_idx, ((a_idx, b_idx), rgb) in enumerate(zip(core, np.random.random((len(core), 3)))):
+        atom_colors_a[int(a_idx)] = tuple(rgb.tolist())
+        atom_colors_b[int(b_idx)] = tuple(rgb.tolist())
+        atom_colors_q[int(c_idx)] = tuple(rgb.tolist())
+
+    if show_idxs:
+        for atom in mol_a_2d.GetAtoms():
+            atom.SetProp("molAtomMapNumber", str(atom.GetIdx()))
+        for atom in mol_b_2d.GetAtoms():
+            atom.SetProp("molAtomMapNumber", str(atom.GetIdx()))
+        for atom in mol_q_2d.GetAtoms():
+            atom.SetProp("molAtomMapNumber", str(atom.GetIdx()))
+
+    return Draw.MolsToGridImage(
+        [mol_q_2d, mol_a_2d, mol_b_2d],
+        molsPerRow=3,
+        highlightAtomLists=[list(range(mol_q_2d.GetNumAtoms())), core[:, 0].tolist(), core[:, 1].tolist()],
+        highlightAtomColors=[atom_colors_q, atom_colors_a, atom_colors_b],
+        subImgSize=(300, 300),
+        legends=["core", mol_a.GetProp("_Name"), mol_b.GetProp("_Name")],
+        useSVG=True,
+    )
+
+
+def sample(bound_impls, masses, x0, v0, box0, lamb, group_idxs, seed, n_frames=500, burn_in=10000):
+
+    dt = 1e-3
+
+    temperature = 300.0
+    friction = 1.0
+
+    intg = LangevinIntegrator(temperature, dt, friction, masses, seed)
+    intg_impl = intg.impl()
+
+    baro = MonteCarloBarostat(len(masses), 1.0, temperature, group_idxs, 15, seed + 1)
+    baro_impl = baro.impl(bound_impls)
+
+    # tbd: add barostat
+    ctxt = custom_ops.Context(x0, v0, box0, intg_impl, bound_impls, baro_impl)
+
+    # burn-in
+    ctxt.multiple_steps_U(
+        lamb=lamb,
+        n_steps=burn_in,
+        lambda_windows=[lamb],
+        store_u_interval=0,
+        store_x_interval=0,
+    )
+
+    steps_per_frame = 750
```

Comment:
Short comment about the choice of 750 here

Reply:
willfix

Reply:
I'm going to set this to 1000 (was what we had for RABFE)